allennlp_models/seq2seq/bart/bart.py

```diff
@@ -0,0 +1,401 @@
+from typing import Dict, Tuple, Any
+from overrides import overrides
+
+from allennlp.data import Vocabulary
+from allennlp.data.fields.text_field import TextFieldTensors
+from allennlp.data.token_indexers.pretrained_transformer_indexer import PretrainedTransformerIndexer
+from allennlp.models.model import Model
+from allennlp.modules.seq2seq_encoders.seq2seq_encoder import Seq2SeqEncoder
+from allennlp.nn.beam_search import BeamSearch
+from allennlp.nn.util import sequence_cross_entropy_with_logits
+from allennlp.training.metrics import BLEU
+from allennlp.training.metrics import ROUGE
+
+from transformers import BartModel, BartForConditionalGeneration
+
+import torch
+from torch import nn
+
+
+@Seq2SeqEncoder.register("bart_encoder")
+class BartEncoder(Seq2SeqEncoder):
+    """
+    The BART encoder without the token and position embeddings.
+    """
+
+    def __init__(self, model_name):
+        super().__init__()
+
+        bart = BartModel.from_pretrained(model_name)
+        self.hidden_dim = bart.config.hidden_size
+        self.bart_encoder = bart.encoder
+        self.bart_encoder.embed_tokens = lambda x: x
+        self.bart_encoder.embed_positions = lambda x: torch.zeros(
+            (x.shape[0], x.shape[1], self.hidden_dim), dtype=torch.float32
+        )
+
+    @overrides
+    def get_input_dim(self) -> int:
+        return self.hidden_dim
+
+    @overrides
+    def get_output_dim(self) -> int:
+        return self.hidden_dim
+
+    @overrides
+    def is_bidirectional(self) -> bool:
+        return False
+
+    @overrides
+    def forward(self, inputs: torch.Tensor, mask: torch.BoolTensor):
+        # The first element is always the last encoder states for each input token.
+        # Depending on the config, the second output will contain a list of the encoder states
+        # after each transformer layer. Similarly, the third output can contain the attentions from each layer.
+        # We only care about the first element.
+        return self.bart_encoder(input_ids=inputs, attention_mask=mask)[0]
+
+
+class _BartEncoderWrapper(nn.Module):
```

Comment:
So `_BartEncoderWrapper` lets you use any AllenNLP seq2seq module as encoder, and `BartEncoder` lets you plug the original BART encoder back in? `b = Bart(..., encoder=None)` functions the same as `b = Bart(..., encoder=BartEncoder())`?

Reply:
Yes, that would work, but is clearly useless. The idea behind BartEncoder is that you can use it as part of a different model.